File: C#/84. Largest Rectangle in Histogram.cs
/*
 * Link: https://leetcode.com/problems/largest-rectangle-in-histogram/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public int LargestRectangleArea(int[] heights) {
        var st = new Stack<int>();
        var f_area = 0;
        var i = 0;
        
        while(i < heights.Length) {
            if (st.Count == 0 || heights[st.Peek()] < heights[i]) {
                st.Push(i++);
            }
            else {
                var height = heights[st.Pop()];
                var width = st.Count > 0 ? i - st.Peek() - 1 : i;
                
                f_area = Math.Max(f_area, height * width);
            }
        }
        
        while(st.Count > 0) {
            var height = heights[st.Pop()];
            var width = st.Count > 0 ? i - st.Peek() - 1: i;

            f_area = Math.Max(f_area, height * width);
        }
        
        return f_area;
    }
}

File: Python/84. Largest Rectangle in Histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        i, res, st = 0, 0, []
        while i < len(heights):
            if len(st) == 0 or heights[st[-1]] < heights[i]:
                st.append(i)
                i += 1 
            else:
                h = heights[st.pop()]
                w = i if len(st) == 0 else i - 1 - st[-1]
                res = max(res, h * w)
        while len(st) > 0:
            h = heights[st.pop()]
            w = i if len(st) == 0 else i - 1 - st[-1]
            res = max(res, h * w)
        return res